switcheroo/data_stores.py

```diff
@@ -0,0 +1,161 @@
+        self._dir: Path = paths.local_ssh_home() if ssh_home is None else ssh_home
+        # To be used later if we decide to use this as a context manager, and temp is selected
+        self._temp_dir: TemporaryDirectory[str] | None = None
+
+    @property
+    def home_dir(self) -> Path:
+        """The folder where local files are stored"""
+        if self.temp and self._temp_dir is not None:
+            return Path(self._temp_dir.name)
+        return self._dir
+
+    @abstractmethod
+    def get_sshd_config_line(self):
+        "Return the config line that the server will add to the sshd_config"
+
+    @abstractmethod
+    def publish(self, host: str, user: str) -> str:
+        """Publish a new public key for the given host and user"""
+
+    @abstractmethod
+    def publish_with_metadata(
+        self, host: str, user: str, metadata: KeyMetadata | None
+    ) -> tuple[str, KeyMetadata]:
+        """Publish a new public key for the given host and user with metadata"""
+
+    @abstractmethod
+    def retrieve(self, host: str, user: str) -> str:
+        """Retrieve the public key for the given host and user"""
+
+    def __enter__(self):
+        if self.temp:
+            self._temp_dir: TemporaryDirectory[str] | None = TemporaryDirectory[
+                str
+            ](  # pylint: disable=consider-using-with
+                prefix="ssh-keys-", dir=self._dir
+            )
+
+    def __exit__(self, exc_t, exc_v, exc_tb):
+        if self.temp:
+            assert self._temp_dir is not None
+            self._temp_dir.__exit__(None, None, None)
+
+
+class S3DataStore(DataStore):
+    "Store the public keys in an S3 bucket"
+
+    def __init__(
+        self, _s3_bucket_name: str, ssh_home: Path | None = None, temp: bool = False
+    ):
+        super().__init__(ssh_home=ssh_home, temp=temp)
+        self._s3_bucket_name = _s3_bucket_name
+
+    @property
+    def s3_bucket_name(self):
+        "The name of the bucket in which the keys are stored"
+        return self._s3_bucket_name
+
+    def get_sshd_config_line(self) -> str:
+        return f"s3 {self.s3_bucket_name}"
+
+    def retrieve(self, host: str, user: str):
+        s3_client = boto3.client("s3")
+        response = s3_client.get_object(
+            Bucket=self.s3_bucket_name, Key=str(paths.cloud_public_key_loc(host, user))
+        )
+        ssh_key = response["Body"].read().decode()
+        return ssh_key
+
+    def publish(self, host: str, user: str) -> str:
+        # Generate new public/private key pair
+        private_key, public_key = KeyGen.generate_private_public_key()
+        # Store the new public key in S3 bucket
+        s3_client = boto3.client("s3")
+        s3_client.put_object(
+            Body=public_key,
+            Bucket=self.s3_bucket_name,
+            Key=str(paths.cloud_public_key_loc(host, user)),
+        )
+
+        # Store the private key on the local machine
+        KeyGen.store_private_key(
```

Comment:
👍

Reply:
While I agree it doesn't make so much sense for key writing/storing logic to be in a class called KeyGen, does it necessarily have to be in the Publisher method itself? It would be helpful to have it as a utility function called by the DataStore for a few reasons:

- The S3Publisher still writes private keys to files - it can just call that utility method (for the private key, at least)
- Unit testing the server - want to avoid using a Publisher. Better to call a utility method

Thanks!